cuqi/solver/_solver.py

```diff
@@ -320,6 +326,137 @@ def solve(self):
         return x, k
 
 
+class PCGLS:
+    """Conjugate Gradient method for least squares problems with preconditioning
+    
+    See Bjorck (1996) - Numerical Methods for Least Squares Problems. Pag 294.
+
+    Parameters
+    ----------
+    A : ndarray or callable f(x,*args).
+    b : ndarray.
+    x0 : ndarray. Initial guess.
+    P : ndarray. Preconditioner array in sparse format.
+    maxit : The maximum number of iterations.
+    tol : The numerical tolerance for convergence checks.
+    """    
+    def __init__(self, A, b, x0, P, maxit, tol=1e-6, shift=0):
+        self.A = A
+        self.b = b
+        self.x0 = x0
+        self.P = P
+        self.maxit = int(maxit)
+        self.tol = tol        
+        self.shift = shift
+        self.dim = len(x0)
+        if not callable(A):
+            self.explicitA = True
+        else:
+            self.explicitA = False
+        #
+        if self.dim < 1000:
+            self.explicitPinv = True
+            Pinv = spa.linalg.inv(P)
+        else:
+            self.explicitPinv = False
+            Pinv = None # we do cholesky.solve or sparse.solve with P
+            if has_cholmod:
+                # turn P into a cholesky object
+                P = cholesky(P, ordering_method='natural')
+        self.Pinv = Pinv # inverse of the preconditioner as a ndarray or function
+
+    def solve(self):
+        # initial state
+        x = self.x0.copy()
+        if self.explicitA:
+            r = self.b - (self.A @ x)
+            if self.explicitPinv:
+                s = self.Pinv.T @ (self.A.T @ r)
+            else:
+                s = self._apply_Pinv((self.A.T @ r), 2)
+        else:
+            r = self.b - self.A(x, 1)
+            if self.explicitPinv:
+                s = self.Pinv.T @ self.A(r, 2)
+            else:
+                s = self._apply_Pinv(self.A(r, 2), 2)
+
+        # initialization
+        p = s.copy()
+        norms0 = LA.norm(s)
+        normx = LA.norm(x)
+        gamma, xmax = norms0**2, normx
+    
+        # main loop
+        k, flag, indefinite = 0, 0, 0
+        while (k < self.maxit) and (flag == 0):
+            k += 1
+            #
+            if self.explicitPinv:
+                t = self.Pinv @ p
+            else:
+                t = self._apply_Pinv(p, 1)
+            if self.explicitA:
+                q = self.A @ t
+            else:
+                q = self.A(t, 1)
```

Comment:
Similar to how you created a function for `_apply_Pinv` I would suggest adding one for `_apply_A`.